File: services/client/src/lottery/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"net"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	Uint32Size   = 4  // bytes de un uint32
	Uint8Size    = 1  // bytes de un uint8
	BirthdateLen = 10 // tamaño de fecha (YYYY-MM-DD)
	HeaderSize   = 5  // 1 byte de tipo + 4 de len de payload

	TypeBets       = 0
	TypeAck        = 1
	TypeNoMoreBets = 2
)

type Bet struct {
	AgencyId  uint32
	FirstName string
	LastName  string
	Document  uint32
	Birthdate string
	Number    uint32
}

type Packet struct {
	typ     byte
	payload []byte
}

func (p *Packet) IsBets() bool       { return p.typ == TypeBets }
func (p *Packet) IsAck() bool        { return p.typ == TypeAck }
func (p *Packet) IsNoMoreBets() bool { return p.typ == TypeNoMoreBets }
func (p *Packet) Payload() []byte    { return p.payload }

func BetsToBytes(bets []Bet, agencyId uint32) []byte {
	buf := make([]byte, 0, Uint32Size*2)
	buf = binary.BigEndian.AppendUint32(buf, uint32(len(bets)))
	buf = binary.BigEndian.AppendUint32(buf, agencyId)
	for _, bet := range bets {
		buf = append(buf, BetToBytes(bet)...)
	}
	return buf
}

func BetToBytes(bet Bet) []byte {
	buf := make([]byte, 0, 60)

	buf = append(buf, byte(len(bet.FirstName)))
	buf = append(buf, bet.FirstName...)

	buf = append(buf, byte(len(bet.LastName)))
	buf = append(buf, bet.LastName...)

	buf = binary.BigEndian.AppendUint32(buf, bet.Document)

	buf = append(buf, bet.Birthdate...)

	buf = binary.BigEndian.AppendUint32(buf, bet.Number)

	return buf
}

func BetsFromBytes(payload []byte) ([]Bet, error) {
	if len(payload) < Uint32Size*2 {
		return nil, fmt.Errorf("payload de bets incompleto")
	}
	count := int(binary.BigEndian.Uint32(payload[:Uint32Size]))
	agencyId := binary.BigEndian.Uint32(payload[Uint32Size : Uint32Size*2])
	pos := Uint32Size * 2
	bets := make([]Bet, 0, count)
	for range count {
		bet, newPos, err := BetFromBytes(payload, pos, agencyId)
		if err != nil {
			return nil, err
		}
		pos = newPos
		bets = append(bets, bet)
	}
	return bets, nil
}

func BetFromBytes(b []byte, pos int, agencyId uint32) (Bet, int, error) {
	firstName, pos, err := readString(b, pos)
	if err != nil {
		return Bet{}, 0, err
	}

	lastName, pos, err := readString(b, pos)
	if err != nil {
		return Bet{}, 0, err
	}

	if pos+Uint32Size > len(b) {
		return Bet{}, 0, fmt.Errorf("payload incompleto")
	}
	document := binary.BigEndian.Uint32(b[pos : pos+Uint32Size])
	pos += Uint32Size

	if pos+BirthdateLen > len(b) {
		return Bet{}, 0, fmt.Errorf("payload incompleto")
	}
	birthday := string(b[pos : pos+BirthdateLen])
	pos += BirthdateLen

	if pos+Uint32Size > len(b) {
		return Bet{}, 0, fmt.Errorf("payload incompleto")
	}
	number := binary.BigEndian.Uint32(b[pos : pos+Uint32Size])
	pos += Uint32Size

	return Bet{
		AgencyId:  agencyId,
		FirstName: firstName,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthday,
		Number:    number,
	}, pos, nil
}

func readString(b []byte, pos int) (string, int, error) {
	if pos+Uint8Size > len(b) {
		return "", 0, fmt.Errorf("payload incompleto")
	}
	strLen := int(b[pos])
	pos += Uint8Size
	if pos+strLen > len(b) {
		return "", 0, fmt.Errorf("payload incompleto")
	}
	str := string(b[pos : pos+strLen])
	pos += strLen
	return str, pos, nil
}

func makePacket(pktType byte, payload []byte) []byte {
	packet := make([]byte, HeaderSize, HeaderSize+len(payload))
	packet[0] = pktType
	binary.BigEndian.PutUint32(packet[1:5], uint32(len(payload)))
	return append(packet, payload...)
}

func MakePacketNoMoreBets() []byte {
	return makePacket(TypeNoMoreBets, nil)
}

func MakePacketAck() []byte {
	return makePacket(TypeAck, nil)
}

func MakePacketBets(bets []Bet) []byte {
	return makePacket(TypeBets, BetsToBytes(bets, bets[0].AgencyId))
}

func ReadMessage(conn net.Conn) (*Packet, error) {
	header, err := safe_socket.RecvAll(conn, HeaderSize)
	if err != nil {
		return nil, err
	}
	msgType := header[0]
	length := binary.BigEndian.Uint32(header[1:5])
	payload, err := safe_socket.RecvAll(conn, int(length))
	if err != nil {
		return nil, err
	}
	return &Packet{typ: msgType, payload: payload}, nil
}

File: services/server/src/protocol/protocol.py
import safe_socket
from lottery.bet import Bet

UINT32_SIZE = 4  # bytes de un uint32
UINT8_SIZE = 1  # bytes de un uint8
BIRTHDATE_LEN = 10  # tamaño de fecha (YYYY-MM-DD)
HEADER_SIZE = 5  # 1 byte de type + 4 de len de payload

TYPE_BETS = 0
TYPE_ACK = 1
TYPE_NO_MORE_BETS = 2


# definición del tipo paquete
class Packet:
    def __init__(self, msg_type: int, payload: bytes):
        self._type = msg_type
        self._payload = payload

    def is_bets(self) -> bool:
        return self._type == TYPE_BETS

    def is_ack(self) -> bool:
        return self._type == TYPE_ACK

    def is_no_more_bets(self) -> bool:
        return self._type == TYPE_NO_MORE_BETS

    def payload(self) -> bytes:
        return self._payload


# serialización
def bets_to_bytes(bets: list[Bet], agency_id: int) -> bytes:
    buf = len(bets).to_bytes(UINT32_SIZE, "big")
    buf += agency_id.to_bytes(UINT32_SIZE, "big")
    for bet in bets:
        buf += bet_to_bytes(bet)
    return buf

def bet_to_bytes(bet: Bet) -> bytes:
    buf = b""

    first = bet.first_name.encode("utf-8")
    buf += len(first).to_bytes(UINT8_SIZE, "big") + first

    last = bet.last_name.encode("utf-8")
    buf += len(last).to_bytes(UINT8_SIZE, "big") + last

    buf += bet.document.to_bytes(UINT32_SIZE, "big")

    buf += bet.birthdate.encode("utf-8")

    buf += bet.number.to_bytes(UINT32_SIZE, "big")
    return buf


def bets_from_bytes(data: bytes) -> list[Bet]:
    count = int.from_bytes(data[:UINT32_SIZE], "big")
    agency_id = int.from_bytes(data[UINT32_SIZE : UINT32_SIZE * 2], "big")
    bets = []
    pos = UINT32_SIZE * 2
    for _ in range(count):
        bet, pos = bet_from_bytes(data, pos, agency_id)
        bets.append(bet)
    return bets


def bet_from_bytes(data: bytes, pos: int, agency_id: int) -> tuple[Bet, int]:
    first_len = int(data[pos])
    pos += UINT8_SIZE
    first_name = data[pos : pos + first_len].decode("utf-8")
    pos += first_len

    last_len = int(data[pos])
    pos += UINT8_SIZE
    last_name = data[pos : pos + last_len].decode("utf-8")
    pos += last_len

    document = int.from_bytes(data[pos : pos + UINT32_SIZE], "big")
    pos += UINT32_SIZE

    birthdate = data[pos : pos + BIRTHDATE_LEN].decode("utf-8")
    pos += BIRTHDATE_LEN

    number = int.from_bytes(data[pos : pos + UINT32_SIZE], "big")
    pos += UINT32_SIZE

    return Bet(agency_id, first_name, last_name, document, birthdate, number), pos


# creadores de paquetes
def make_packet(msg_type: int, payload: bytes = b"") -> bytes:
    return bytes([msg_type]) + len(payload).to_bytes(UINT32_SIZE, "big") + payload


def make_packet_no_more_bets() -> bytes:
    return make_packet(TYPE_NO_MORE_BETS)


def make_packet_ack() -> bytes:
    return make_packet(TYPE_ACK)


def make_packet_bets(bets: list[Bet], agency_id: int) -> bytes:
    return make_packet(TYPE_BETS, bets_to_bytes(bets, agency_id))


# leer del socket para formar un paquete
def read_message(sock) -> Packet:
    header = safe_socket.recv_all(sock, HEADER_SIZE)
    msg_type = header[0]
    length = int.from_bytes(header[1:HEADER_SIZE], "big")
    payload = safe_socket.recv_all(sock, length) if length else b""
    return Packet(msg_type, payload)
